module/os/windows/spl/spl-kmem.c

```diff
@@ -0,0 +1,6628 @@
+	 * update our contention count.  We use the depot
+	 * contention rate to determine whether we need to
+	 * increase the magazine size for better scalability.
+	 */
+	if (!mutex_tryenter(&cp->cache_depot_lock)) {
+		mutex_enter(&cp->cache_depot_lock);
+		cp->cache_depot_contention++;
+	}
+
+	if ((mp = mlp->ml_list) != NULL) {
+		ASSERT(KMEM_MAGAZINE_VALID(cp, mp));
+		mlp->ml_list = mp->mag_next;
+		if (--mlp->ml_total < mlp->ml_min)
+			mlp->ml_min = mlp->ml_total;
+		mlp->ml_alloc++;
+	}
+
+	mutex_exit(&cp->cache_depot_lock);
+
+	return (mp);
+}
+
+/*
+ * Free a magazine to the depot.
+ */
+static void
+kmem_depot_free(kmem_cache_t *cp, kmem_maglist_t *mlp, kmem_magazine_t *mp)
+{
+	mutex_enter(&cp->cache_depot_lock);
+	ASSERT(KMEM_MAGAZINE_VALID(cp, mp));
+	mp->mag_next = mlp->ml_list;
+	mlp->ml_list = mp;
+	mlp->ml_total++;
+	mutex_exit(&cp->cache_depot_lock);
+}
+
+/*
+ * Update the working set statistics for cp's depot.
+ */
+static void
+kmem_depot_ws_update(kmem_cache_t *cp)
+{
+	mutex_enter(&cp->cache_depot_lock);
+	cp->cache_full.ml_reaplimit = cp->cache_full.ml_min;
+	cp->cache_full.ml_min = cp->cache_full.ml_total;
+	cp->cache_empty.ml_reaplimit = cp->cache_empty.ml_min;
+	cp->cache_empty.ml_min = cp->cache_empty.ml_total;
+	mutex_exit(&cp->cache_depot_lock);
+}
+
+/*
+ * Set the working set statistics for cp's depot to zero. (Everything is
+ * eligible for reaping.)
+ */
+void
+kmem_depot_ws_zero(kmem_cache_t *cp)
+{
+	mutex_enter(&cp->cache_depot_lock);
+	cp->cache_full.ml_reaplimit = cp->cache_full.ml_total;
+	cp->cache_full.ml_min = cp->cache_full.ml_total;
+	cp->cache_empty.ml_reaplimit = cp->cache_empty.ml_total;
+	cp->cache_empty.ml_min = cp->cache_empty.ml_total;
+	mutex_exit(&cp->cache_depot_lock);
+}
+
+/*
+ * The number of bytes to reap before we call kpreempt(). The default (1MB)
+ * causes us to preempt reaping up to hundres of times per second.  Using a
+ * larger value (1GB) causes this to have virtually no effect.
+ */
+size_t kmem_reap_preempt_bytes = 64 * 1024 * 1024;
+
+
+/*
+ * Reap all magazines that have fallen out of the depot's working set.
+ */
+static void
+kmem_depot_ws_reap(kmem_cache_t *cp)
+{
+	size_t bytes = 0;
+	long reap;
+	kmem_magazine_t *mp;
+
+	ASSERT(!list_link_active(&cp->cache_link) ||
+	    taskq_member(kmem_taskq, curthread));
+
+	reap = MIN(cp->cache_full.ml_reaplimit, cp->cache_full.ml_min);
+	while (reap-- &&
+	    (mp = kmem_depot_alloc(cp, &cp->cache_full)) != NULL) {
+		kmem_magazine_destroy(cp, mp, cp->cache_magtype->mt_magsize);
+		bytes += cp->cache_magtype->mt_magsize * cp->cache_bufsize;
+		if (bytes > kmem_reap_preempt_bytes) {
+			kpreempt(KPREEMPT_SYNC);
+			bytes = 0;
+		}
+	}
+
+	reap = MIN(cp->cache_empty.ml_reaplimit, cp->cache_empty.ml_min);
+	while (reap-- &&
+	    (mp = kmem_depot_alloc(cp, &cp->cache_empty)) != NULL) {
+		kmem_magazine_destroy(cp, mp, 0);
+		bytes += cp->cache_magtype->mt_magsize * cp->cache_bufsize;
```

Comment:
This is probably okay, but we should either add a defensive `size_t` cast, or change the types of  `->cache_bufsize` and `->cache_align` to `size_t` to match `umem_cache_t`.